Add Crossdart links

astashov opened the pull request:
Added them to the _source_code partial, so every time we show the source code for something, it will have the Crossdart link.

/cc @sethladd

astashov commented:
I signed it!

sethladd commented:
Cool!

Can you add a command-line option to toggle this?

astashov commented:
Sure!

sethladd commented on lib/src/model.dart:
```diff
@@ -1268,8 +1271,46 @@ abstract class SourceCodeMixin {
+    if (_lineNumber != null && _sourceFilePath != null) {
+      String packageName = library.package.isSdk ? "sdk" : library.package.name;
+      String packageVersion = library.package.version;
+      String url = path.join("http://www.crossdart.info/p", packageName, packageVersion, _sourceFilePath) + ".html";
```
we would like this to be https or relative (`//www.crossdart`)

astashov replied:
Will do `//www.crossdart`, thanks

sethladd replied:
can you host crossdart on SSL?

astashov replied:
Sure, I host it on Google Storage (it's a static site), so I just need to buy SSL certificate, that's it.

sethladd replied:
Unfortunately, google cloud storage doesn't allow for custom certs + custom DNS :(

You could put cloudfront in front.

astashov replied:
Oh really? Okay, I need to look into it then. Thanks for advice.

astashov commented:
Okay, so, crossdart.info now supports HTTPS (https://crossdart.info). Unfortunately, I couldn't figure out how to make it work with CloudFront and GCS, so I moved everything to S3+CloudFront+Route53.

Also, added a CLI flag to turn Crossdart on.

@sethladd, PTAL.

sethladd commented:
Hi!

OK, we got our build in order and tests should be passing on our buildbots. Please take a moment to rebase and send up a new diff. Thanks!

sethladd commented on lib/src/cache.dart:
```diff
@@ -0,0 +1,55 @@
+  }
+}
+
+var cache = new Cache();
```
Can you use a type annotation here? top-level variables should be typed. Thanks!

astashov commented:
Rebased. It's weird - Travis CI complains that I have to run `dartfmt -w lib/src/model.dart`, but when I run it, it doesn't make any changes. Any idea why? Maybe I miss some dartfmt config file or something like that?

devoncarew commented:
Travis is running `1.12.0-dev.5.10` for that build. I'm guessing that the dev formatter produces slightly different code than the stable formatter. Are you on stable? That could be it -

sethladd commented:
What @devoncarew said. dartfmt changed its style between releases.

Now that 1.12 is out, we'll turn on stable-channel testing again and everyone can and should be using 1.12's dartfmt.

Sorry for the issues!

astashov commented:
Yup, just noticed, already upgrading our projects to Dart 1.12.0. Null-aware operators FTW! :)
Will check this one out tonight.

astashov commented:
Okay, Travis is now green, I squashed everything into one commit, and ready to go! :)

sethladd commented on lib/src/model.dart:
```diff
@@ -1277,8 +1283,49 @@ abstract class SourceCodeMixin {
+    }
+  }
+
+  String get _crossdartUrl {
```
Sorry I missed this before. Can you add some tests for this?

We don't need to test the generated HTML but it would be nice to test the output of calling `element.documentation` and `element.documentationAsHtml` with this option on and off.

astashov replied:
Makes sense, will add.

sethladd replied:
Thanks. Apologies for not catching that before.

astashov replied:
No problem :) Added tests, PTAL.

sethladd commented:
Nice! lgtm.

I'll merge this in for you.

astashov commented:
Okay, let me try that.

astashov commented:
All green now, PTAL, @sethladd!

astashov commented:
Ping? :)

sethladd commented:
Ack, sorry I missed this. Looks like you need to rebase. And then we'll merge it.

sethladd commented:
Friendly ping. Care to rebase this so we can merge? Thanks!

astashov commented:
Rebased! Sorry for delay.

sethladd commented:
Super cool, thanks!
